File: matlab/utilities/basis/eigh_descending_sym.m
function [evals_sorted, evecs_sorted] = eigh_descending_sym(matrix, do_symmetrize)
% EIGH_DESCENDING_SYM  Compute eigendecomposition with consistent sorting
%
%   [evals_sorted, evecs_sorted] = eigh_descending_sym(matrix) computes
%   the eigendecomposition of a symmetric matrix and returns eigenvalues
%   and eigenvectors sorted by descending eigenvalue magnitude, with
%   standardized eigenvector signs for reproducibility.
%
%   [evals_sorted, evecs_sorted] = eigh_descending_sym(matrix, do_symmetrize)
%   optionally forces the matrix to be symmetric before eigendecomposition.
%
% -------------------------------------------------------------------------
% Inputs:
% -------------------------------------------------------------------------
%
% <matrix> - [n x n] numeric matrix (should be symmetric or nearly symmetric)
%
% <do_symmetrize> (optional) - logical. If true, enforces symmetry via
%   (matrix + matrix')/2 before eigendecomposition. Default: false.
%   Note: GSN returns symmetric cSb/cNb, so symmetrization is typically
%   only needed for derived matrices like cSb - cNb/ntrials_avg
%
% -------------------------------------------------------------------------
% Returns:
% -------------------------------------------------------------------------
%
% <evals_sorted> - [n x 1] eigenvalues sorted in descending order
%
% <evecs_sorted> - [n x n] eigenvectors with columns sorted to match
%   <evals_sorted>. Signs standardized so that the largest-magnitude
%   element in each column is positive

    if nargin < 2
        do_symmetrize = false; %#ok<NASGU>  % retained for API compatibility
    end

    % Always symmetrize. This function is for symmetric matrices, and MATLAB's
    % general eig() returns COMPLEX eigenvectors when given a matrix with even
    % machine-epsilon asymmetry (which GSN covariances can have). Python's
    % np.linalg.eigh is always a symmetric solver, so symmetrizing here keeps the
    % results real and matches Python. For an exactly-symmetric matrix this is a
    % no-op, so it does not change any existing result.
    matrix = (matrix + matrix') / 2;

    % Compute eigendecomposition
    [evecs, evals] = eig(matrix, 'vector');

    % Sort by eigenvalue magnitude (descending)
    [evals_sorted, order] = sort(evals, 'descend');
    evecs_sorted = evecs(:, order);

    % Deterministic sign: make largest-magnitude element positive
    [~, piv] = max(abs(evecs_sorted), [], 1);
    idx = sub2ind(size(evecs_sorted), piv, 1:size(evecs_sorted, 2));
    sgn = sign(evecs_sorted(idx));
    sgn(sgn == 0) = 1;
    evecs_sorted = evecs_sorted .* sgn;
end

File: psn/utilities/basis/eigh_descending_sym.py
"""Eigendecomposition utility with consistent sorting."""

import numpy as np

from psn._device import from_device, is_cpu, resolve_device, to_device


def eigh_descending_sym(matrix, do_symmetrize=False, device='cpu'):
    """Compute eigendecomposition with consistent sorting.

    Computes the eigendecomposition of a symmetric matrix and returns
    eigenvalues and eigenvectors sorted by descending eigenvalue magnitude,
    with standardized eigenvector signs for reproducibility.

    Parameters
    ----------
    matrix : ndarray, shape (n, n)
        Numeric matrix (should be symmetric or nearly symmetric).
    do_symmetrize : bool, optional
        If True, enforces symmetry via (matrix + matrix.T)/2 before
        eigendecomposition. Default: False.
        Note: GSN returns symmetric cSb/cNb, so symmetrization is typically
        only needed for derived matrices like cSb - cNb/ntrials_avg.
    device : str or torch.device, optional
        Where to run the O(n^3) eigh. 'cpu' (default) uses numpy. A GPU device
        ('cuda'/'mps') runs it via torch on-device. Note: a GPU eigh picks a
        different orthonormal basis on degenerate eigenspaces than numpy, so
        downstream thresholds can differ by a few percent from a CPU run - an
        inherent property of the backend, not a bug.

    Returns
    -------
    evals_sorted : ndarray, shape (n,)
        Eigenvalues sorted in descending order.
    evecs_sorted : ndarray, shape (n, n)
        Eigenvectors with columns sorted to match evals_sorted.
        Signs standardized so that the largest-magnitude element
        in each column is positive.
    """
    if do_symmetrize:
        matrix = (matrix + matrix.T) / 2

    # Compute eigendecomposition on the requested device.
    dev = resolve_device(device)
    if is_cpu(dev):
        evals, evecs = np.linalg.eigh(matrix)
    else:
        import torch
        tdtype = (torch.float64 if np.asarray(matrix).dtype == np.float64
                  else torch.float32)
        M_t = to_device(matrix, dev, dtype=tdtype)
        evals_t, evecs_t = torch.linalg.eigh(M_t)
        evals = from_device(evals_t)
        evecs = from_device(evecs_t)
        del M_t, evals_t, evecs_t

    # Sort by eigenvalue magnitude (descending)
    order = np.argsort(evals)[::-1]
    evals_sorted = evals[order]
    evecs_sorted = evecs[:, order]

    # Deterministic sign: make largest-magnitude element positive
    piv = np.argmax(np.abs(evecs_sorted), axis=0)
    idx = (piv, np.arange(evecs_sorted.shape[1]))
    sgn = np.sign(evecs_sorted[idx])
    sgn[sgn == 0] = 1
    evecs_sorted = evecs_sorted * sgn

    return evals_sorted, evecs_sorted
